### DISCUSSION TITLE
Low Bandwidth Mode

### DISCUSSION THREAD

**Discussion by seanbts** [sinamics/ztnet#310] [Q&A]
Hi,
Is it possible to set the low bandwidth mode in the controller? And if so were would this command be done? Thanks
"lowBandwidthMode":true

**Comment by seanbts** [comment]
Hi Guys,

would anyone have an answer on this?
Thanks and appreciated.

**Comment by n9yty** [comment]
I think this would need to be set on all client configurations. I'm not sure if the controller alone would handle it, or what happens if you didn't configure the controller in this way at all and had all clients configured with it. I don't see that documented, as they focus on "a device"...
 
I don't know if this helps or not.  That setting is set in the local.conf file of the zerotier docker container. It may not yet be settable by the API or zerotier-cli, so ztnet may not be able to change it.  In my case, that zerotier configuration is stored in the docker volume ztnet_zerotier.  Use ```docker volume ls``` to see a list in your case.

In the docker-compose file for ztnet, the zerotier container has the environment variable ZT_OVERRIDE_LOCAL_CONF set to true. By looking at the zyclonite/zerotier container it seems that if this is the case it creates one and overwrites your local one.  If this flag is false, and no local.conf exists, one will be created also.  So if you were going to manually create or alter this you probably want to unset that ZT_OVERRIDE_LOCAL_CONF and be sure to include the equivalent to what is autocreated:

```
{
    "settings": {
        "primaryPort": ${ZT_PRIMARY_PORT:-$DEFAULT_PRIMARY_PORT},
        "portMappingEnabled": ${ZT_PORT_MAPPING_ENABLED:-$DEFAULT_PORT_MAPPING_ENABLED},
        "softwareUpdate": "disable",
        "allowManagementFrom": [${MANAGEMENT_NETWORKS}],
        "allowTcpFallbackRelay": ${ZT_ALLOW_TCP_FALLBACK_RELAY:-$DEFAULT_ALLOW_TCP_FALLBACK_RELAY}
    }
}
```

The local.conf file would be in the docker volume "zerotier". I also see one under networks.d/{networkid}.local.conf, so I'm not sure which is needed here.  Looking at mine, the one in networks.d are settings like "allowManaged" and so forth which are set in the controller GUI and probably not relevant. So you probaby need the one at the top level only.

You can see the current config in a similar manner as to what I will suggest using to edit it:

```
docker run --rm -it -v ztnet_zerotier:/zt ubuntu cat /zt/local.conf
```

A hacky way to get in there to edit it though would be by launching a temporary ubuntu container, installing vi, and editing that file:

```
docker run --rm -it -v ztnet_zerotier:/zt ubuntu bash -c "apt update && apt install vim -y && vi /zt/local.conf"
```

Then you could add the required lowBandwidth as they document:

```
{
  "settings":
  {
    "lowBandwidthMode":true
  }
}
```

Are there better ways? I'm sure there are. But maybe this will help anyway. :)

**Comment by sinamics** [comment]
>I think this would need to be set on all client configurations. I'm not sure if the controller alone would handle it, or what happens if you didn't configure the controller in this way at all and had all clients configured with it. I don't see that documented, as they focus on "a device"...

Im not sure either, nor am i familiar with this parameter.
However, if its enough to just set this on the controller, i can add an option to set `lowBandwidthMode`.

> It may not yet be settable by the API or zerotier-cli, so ztnet may not be able to change it

Ztnet is already making chagnes to the local.conf file, so adding an option to set `lowBandwidthMode` is quite trivial.



> Are there better ways?

You can just open an editor directly in the zerotier folder if you need to edit config files.
example:
`docker exec -it ztnet nano /var/lib/zerotier-one/local.conf`

**Comment by n9yty** [comment]
I should have linked to the documentation, but there isn’t much:

https://docs.zerotier.com/lbm/

**Comment by sinamics** [comment]
https://www.reddit.com/r/zerotier/comments/1988664/comment/ki5lfpg/?utm_source=share&utm_medium=web3x&utm_name=web3xcss&utm_term=1&utm_content=share_button

**Comment by n9yty** [comment]
I changed the local.conf for all the nodes on our network, and on the ZeroTier hosted controller all of the online status information went away. They all show offline. On the ZTNet controller status page they come and go from online to offilne and many show RELAYED. I did change the local.conf on the controller docker container as well. Maybe that should have been left alone? Just trying to get a feel for it, but I obviously did not and could not change the ZeroTier hosted controller config. :)

**Comment by sinamics** [comment]
Interesting.
I'm curious about the impact on peer connections when operating in `lowBandwidthMode`. 
To monitor the behavior of peers, you can use the command `watch docker exec -it zerotier zerotier-cli listpeers`


I will test myself when i get time.

**Reply by n9yty** [reply]
I did notice some odd behavior, so I have already backed this out of our deployment. Interesting, it seems that you can not set that flag to false ("lowBandwidthMode": false) because then the zerotier network didn't start up at all. i had to remove it. :) We did not really have a need for that mode, so it wasn't worth the odd little issues like nodes being slow to respond, even though I don't think they had any network changes, lack of visibility in the network controller, etc.

**Reply by n9yty** [reply]
I wanted to add to this report - I am still seeing odd network issues. I am not sure if it is because all the nodes are on two simultaneous ZT networks, although I don't recall that being a problem at first, or if it is because of something else. I had backed out the lowBandwidth settings but yet connectivity between my computer and the other nodes comes and goes, usually dropping the ZT network connections and restarting them puts things right again. This is on macOS Sonoma 14.4. I am not sure exactly what is going on, but it may or may not mean there was an issue with lowBandwidth mode.
